RFCs/0009-Fn::ForEach.md

```diff
@@ -0,0 +1,1461 @@
+    Properties:
+      ImageId: ami-id3
+      InstanceType: m5.2xlarge
+      AvailabilityZone: us-east-1a
+Resources:
+  'Fn::ForEach::Instances':
+    - InstanceLogicalId
+    - [InstanceA, InstanceB, InstanceC]
+    - '${InstanceLogicalId}':
+        Type: 'AWS::EC2::Instance'
+        Properties:
+          DisableApiTermination: true
+          UserData:
+            Fn::Base64: 
+              !Sub |
+                #!/bin/bash
+                yum update -y
+                yum install -y httpd.x86_64
+                systemctl start httpd.service
+                systemctl enable httpd.service
+                echo "Hello World from $(hostname -f)" > /var/www/html/index.html
+          'Fn::ForEach::Properties':
+            - PropertyName
+            - [ImageId, InstanceType, AvailabilityZone]
+            - '${PropertyName}': !FindInMap [!Ref InstanceLogicalId, Properties, !Ref PropertyName, DefaultValue: !Ref 'AWS::NoValue']
+```
+
+#### The above templates would be equivalent to the below template in YAML
+
+```yaml
+AWSTemplateFormatVersion: 2010-09-09
+Transform: 'AWS::LanguageExtensions'
+Resources:
+  InstanceA:
+    Type: 'AWS::EC2::Instance'
+    Properties:
+      DisableApiTermination: true
+      UserData:
+        Fn::Base64: 
+          !Sub |
+            #!/bin/bash
+            yum update -y
+            yum install -y httpd.x86_64
+            systemctl start httpd.service
+            systemctl enable httpd.service
+            echo "Hello World from $(hostname -f)" > /var/www/html/index.html
+      ImageId: ami-id1
+      InstanceType: m5.xlarge
+  InstanceB:
+    Type: 'AWS::EC2::Instance'
+    Properties:
+      DisableApiTermination: true
+      UserData:
+        Fn::Base64: 
+          !Sub |
+            #!/bin/bash
+            yum update -y
+            yum install -y httpd.x86_64
+            systemctl start httpd.service
+            systemctl enable httpd.service
+            echo "Hello World from $(hostname -f)" > /var/www/html/index.html
+      ImageId: ami-id2
+  InstanceC:
+    Type: 'AWS::EC2::Instance'
+    Properties:
+      DisableApiTermination: true
+      UserData:
+        Fn::Base64: 
+          !Sub |
+            #!/bin/bash
+            yum update -y
+            yum install -y httpd.x86_64
+            systemctl start httpd.service
+            systemctl enable httpd.service
+            echo "Hello World from $(hostname -f)" > /var/www/html/index.html
+      ImageId: ami-id3
+      InstanceType: m5.2xlarge
+      AvailabilityZone: us-east-1a
+```
+
+### Fn::ForEach in Outputs section
+
+#### Use case 5: Reference Replicated Resources
+
+In this example, customer is using two nested `Fn::ForEach` loops in Outputs section to reduce template length.
+
+#### JSON
+
+```json
+{
+    "AWSTemplateFormatVersion": "2010-09-09",
+    "Transform": "AWS::LanguageExtensions",
+    "Resources": {
+        "Fn::ForEach::Buckets": [
+            "Identifier",
+            ["A", "B", "C"],
```

Comment:
Perhaps this example might be better with the collection from a Mapping, so that it's not duplicated between the two?

Reply:
Yeah that would be better, so I’ve updated the example accordingly